File: day06/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
	gridEnd     = 720
	gridStart   = -360
)

type (
	Point struct {
		X, Y int
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	points := []Point{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var x, y int
		fmt.Sscanf(scanner.Text(), "%d, %d", &x, &y)
		points = append(points, Point{
			X: x,
			Y: y,
		})
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	inRegion := 0
	edge := map[int]struct{}{}
	counts := make([]int, len(points))
	for i := gridStart; i < gridEnd; i++ {
		for j := gridStart; j < gridEnd; j++ {
			k, tie := findClosest(j, i, points)
			if !tie {
				counts[k]++
			}
			if isEdge(j, i) {
				edge[k] = struct{}{}
			}
			if combinedDistance(j, i, points) < 10000 {
				inRegion++
			}
		}
	}

	max := 0
	for n, i := range counts {
		if _, ok := edge[n]; !ok && i > max {
			max = i
		}
	}

	fmt.Println(max)
	fmt.Println(inRegion)
}

func combinedDistance(x, y int, points []Point) int {
	p := Point{
		X: x,
		Y: y,
	}
	dist := 0
	for _, i := range points {
		dist += distance(&p, &i)
	}
	return dist
}

func isEdge(x, y int) bool {
	return x == gridStart || y == gridStart || x == gridEnd-1 || y == gridEnd-1
}

func findClosest(x, y int, points []Point) (int, bool) {
	p := Point{
		X: x,
		Y: y,
	}
	tie := false
	ind := 0
	dist := distance(&p, &points[0])
	for n, i := range points[1:] {
		k := distance(&p, &i)
		if k < dist {
			tie = false
			dist = k
			ind = n + 1
		} else if k == dist {
			tie = true
		}
	}
	return ind, tie
}

func distance(a, b *Point) int {
	return abs(a.X-b.X) + abs(a.Y-b.Y)
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

File: day06/src/main.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";
const GRID_START: isize = -1;
const GRID_END: isize = 361;
const DIST_CAP: isize = 10000;

struct Point(isize, isize);

fn distance(a: &Point, b: &Point) -> isize {
    let Point(ax, ay) = a;
    let Point(bx, by) = b;
    (ax - bx).abs() + (ay - by).abs()
}

fn find_closest(p: &Point, points: &Vec<Point>) -> (bool, usize, isize) {
    points.iter().enumerate().skip(1).fold(
        (false, 0, distance(p, &points[0])),
        |(tie, ind, dist), (n, i)| {
            let k = distance(p, i);
            if k < dist {
                (false, n, k)
            } else {
                (k == dist || tie, ind, dist)
            }
        },
    )
}

fn is_edge(p: &Point) -> bool {
    let Point(x, y) = p;
    *x == GRID_START || *y == GRID_START || *x == GRID_END - 1 || *y == GRID_END - 1
}

fn combined_distance(p: &Point, points: &Vec<Point>) -> isize {
    points.iter().fold(0, |acc, i| acc + distance(p, i))
}

fn main() {
    let file = File::open(PUZZLEINPUT).expect("Failed to open file");
    let reader = BufReader::new(file);

    let points = reader
        .lines()
        .map(|x| {
            let k = x
                .expect("Failed to read line")
                .split(", ")
                .map(|x| x.parse::<isize>().expect("Failed to parse num"))
                .collect::<Vec<_>>();
            Point(k[0], k[1])
        })
        .collect::<Vec<_>>();

    let mut in_region = 0;
    let mut edge = HashSet::new();
    let mut counts = vec![0; points.len()];
    for i in GRID_START..GRID_END {
        for j in GRID_START..GRID_END {
            let p = Point(j, i);
            let (tie, ind, _) = find_closest(&p, &points);
            if !tie {
                counts[ind] += 1;
                if is_edge(&p) {
                    edge.insert(ind);
                }
            }
            if combined_distance(&p, &points) < DIST_CAP {
                in_region += 1;
            }
        }
    }

    println!(
        "{}",
        counts
            .iter()
            .enumerate()
            .filter(|(n, _)| !edge.contains(n))
            .max_by_key(|(_, &x)| x)
            .expect("Cannot find max")
            .1
    );

    println!("{}", in_region);
}
